File: pkg/templates/typescript/cua/providers/openai.ts
/**
 * OpenAI CUA provider adapter.
 *
 * Uses the OpenAI Responses API with computer use tool.
 */

import OpenAI from 'openai';
import type {
  ResponseInputItem,
  ResponseItem,
  ResponseComputerToolCall,
  ResponseOutputMessage,
} from 'openai/resources/responses/responses';
import type { CuaProvider, TaskOptions, TaskResult } from './index';

const KEYSYM_MAP: Record<string, string> = {
  ENTER: 'Return', Enter: 'Return', RETURN: 'Return',
  BACKSPACE: 'BackSpace', Backspace: 'BackSpace',
  DELETE: 'Delete', TAB: 'Tab', ESCAPE: 'Escape', Escape: 'Escape',
  SPACE: 'space', Space: 'space',
  UP: 'Up', DOWN: 'Down', LEFT: 'Left', RIGHT: 'Right',
  HOME: 'Home', END: 'End',
  PAGEUP: 'Prior', PAGE_UP: 'Prior', PageUp: 'Prior',
  PAGEDOWN: 'Next', PAGE_DOWN: 'Next', PageDown: 'Next',
  CTRL: 'Control_L', Ctrl: 'Control_L', CONTROL: 'Control_L', Control: 'Control_L',
  ALT: 'Alt_L', Alt: 'Alt_L',
  SHIFT: 'Shift_L', Shift: 'Shift_L',
  META: 'Super_L', Meta: 'Super_L', CMD: 'Super_L', COMMAND: 'Super_L',
  F1: 'F1', F2: 'F2', F3: 'F3', F4: 'F4', F5: 'F5', F6: 'F6',
  F7: 'F7', F8: 'F8', F9: 'F9', F10: 'F10', F11: 'F11', F12: 'F12',
};

const MODIFIER_KEYSYMS = new Set([
  'Control_L', 'Control_R', 'Alt_L', 'Alt_R',
  'Shift_L', 'Shift_R', 'Super_L', 'Super_R',
]);

function translateKeys(keys: string[]): string[] {
  return keys.map(k => KEYSYM_MAP[k] ?? k);
}

function expandAndTranslateKeys(keys: string[], holdKeys: string[]): { keys: string[]; holdKeys: string[] } {
  const expanded: string[] = [];
  for (const raw of keys) {
    const parts = raw.includes('+') ? raw.split('+') : [raw];
    for (const part of parts) {
      const trimmed = part.trim();
      if (trimmed) expanded.push(trimmed);
    }
  }

  const translated = translateKeys(expanded);
  const translatedHold = translateKeys(holdKeys);

  const holdFromKeys: string[] = [];
  const primaryKeys: string[] = [];
  for (const key of translated) {
    if (MODIFIER_KEYSYMS.has(key)) holdFromKeys.push(key);
    else primaryKeys.push(key);
  }

  if (primaryKeys.length === 0) return { keys: translated, holdKeys: translatedHold };

  const merged = [...new Set([...translatedHold, ...holdFromKeys])];
  return { keys: primaryKeys, holdKeys: merged };
}

interface CuaAction {
  type: string;
  x?: number;
  y?: number;
  text?: string;
  url?: string;
  keys?: string[];
  hold_keys?: string[];
  button?: string | number;
  scroll_x?: number;
  scroll_y?: number;
  ms?: number;
  path?: Array<{ x: number; y: number }>;
  [key: string]: unknown;
}

type BatchAction = {
  type: string;
  click_mouse?: { x: number; y: number; button?: string; num_clicks?: number };
  move_mouse?: { x: number; y: number };
  type_text?: { text: string };
  press_key?: { keys: string[]; hold_keys?: string[] };
  scroll?: { x: number; y: number; delta_x?: number; delta_y?: number };
  drag_mouse?: { path: number[][] };
  sleep?: { duration_ms: number };
};

function normalizeButton(button?: string | number): string {
  if (button === undefined || button === null) return 'left';
  if (typeof button === 'number') return button === 2 ? 'middle' : button === 3 ? 'right' : 'left';
  return button;
}

function translateCuaAction(action: CuaAction): BatchAction[] {
  switch (action.type) {
    case 'click': {
      if (action.button === 'back') return [{ type: 'press_key', press_key: { hold_keys: ['Alt_L'], keys: ['Left'] } }];
      if (action.button === 'forward') return [{ type: 'press_key', press_key: { hold_keys: ['Alt_L'], keys: ['Right'] } }];
      if (action.button === 'wheel') {
        return [{ type: 'scroll', scroll: { x: action.x ?? 0, y: action.y ?? 0, delta_x: action.scroll_x ?? 0, delta_y: action.scroll_y ?? 0 } }];
      }
      return [{ type: 'click_mouse', click_mouse: { x: action.x ?? 0, y: action.y ?? 0, button: normalizeButton(action.button) } }];
    }
    case 'double_click':
      return [{ type: 'click_mouse', click_mouse: { x: action.x ?? 0, y: action.y ?? 0, num_clicks: 2 } }];
    case 'type':
      return [{ type: 'type_text', type_text: { text: action.text ?? '' } }];
    case 'keypress': {
      const n = expandAndTranslateKeys(action.keys ?? [], action.hold_keys ?? []);
      return [{ type: 'press_key', press_key: { keys: n.keys, ...(n.holdKeys.length ? { hold_keys: n.holdKeys } : {}) } }];
    }
    case 'scroll':
      return [{ type: 'scroll', scroll: { x: action.x ?? 0, y: action.y ?? 0, delta_x: action.scroll_x ?? 0, delta_y: action.scroll_y ?? 0 } }];
    case 'move':
      return [{ type: 'move_mouse', move_mouse: { x: action.x ?? 0, y: action.y ?? 0 } }];
    case 'drag': {
      const points = (action.path ?? []).map(p => [p.x, p.y]);
      if (points.length < 2) throw new Error('drag requires at least 2 path points');
      return [{ type: 'drag_mouse', drag_mouse: { path: points } }];
    }
    case 'wait':
      return [{ type: 'sleep', sleep: { duration_ms: action.ms ?? 1000 } }];
    case 'goto':
      return [
        { type: 'press_key', press_key: { keys: ['l'], hold_keys: ['Control_L'] } },
        { type: 'sleep', sleep: { duration_ms: 200 } },
        { type: 'press_key', press_key: { keys: ['a'], hold_keys: ['Control_L'] } },
        { type: 'type_text', type_text: { text: action.url ?? '' } },
        { type: 'press_key', press_key: { keys: ['Return'] } },
      ];
    case 'back':
      return [{ type: 'press_key', press_key: { keys: ['Left'], hold_keys: ['Alt_L'] } }];
    case 'screenshot':
      return [];
    default:
      throw new Error(`Unknown CUA action: ${action.type}`);
  }
}

export class OpenAIProvider implements CuaProvider {
  readonly name = 'openai';
  private apiKey: string;

  constructor() {
    this.apiKey = process.env.OPENAI_API_KEY ?? '';
  }

  isConfigured(): boolean {
    return this.apiKey.length > 0;
  }

  async runTask(options: TaskOptions): Promise<TaskResult> {
    const { query, kernel, sessionId } = options;
    const client = new OpenAI({ apiKey: this.apiKey });

    // Navigate to a neutral starting page
    await kernel.browsers.computer.batch(sessionId, {
      actions: translateCuaAction({ type: 'goto', url: 'https://duckduckgo.com' }) as Parameters<typeof kernel.browsers.computer.batch>[1]['actions'],
    });

    const input: ResponseInputItem[] = [
      {
        role: 'system',
        content: `Current date: ${new Date().toISOString()}. For long pages, prefer PageUp/PageDown style scrolling over repeated mouse-wheel scrolling. Use wheel scrolling mainly for small adjustments.`,
      } as unknown as ResponseInputItem,
      {
        type: 'message',
        role: 'user',
        content: [{ type: 'input_text', text: query }],
      },
    ];

    const items: ResponseItem[] = [];
    const maxTurns = 50;

    for (let turn = 0; turn < maxTurns; turn++) {
      const response = await client.responses.create({
        model: options.model || 'gpt-5.4',
        input: [...input, ...items] as ResponseInputItem[],
        tools: [{ type: 'computer' } as unknown as OpenAI.Responses.Tool],
        truncation: 'auto',
        reasoning: { effort: 'low', summary: 'concise' },
      });

      if (!response.output) throw new Error('No output from model');

      for (const item of response.output as ResponseItem[]) {
        items.push(item);

        if (item.type === 'computer_call') {
          const cc = item as ResponseComputerToolCall & {
            action?: CuaAction;
            actions?: CuaAction[];
          };
          const actionList: CuaAction[] = Array.isArray(cc.actions)
            ? cc.actions
            : cc.action ? [cc.action] : [];

          // Execute actions
          const batch: BatchAction[] = [];
          for (const a of actionList) {
            batch.push(...translateCuaAction(a));
          }
          if (batch.length > 0) {
            await kernel.browsers.computer.batch(sessionId, {
              actions: batch as Parameters<typeof kernel.browsers.computer.batch>[1]['actions'],
            });
          }

          // Acknowledge safety checks
          const pending = cc.pending_safety_checks ?? [];
          for (const check of pending) {
            console.log(`Safety check: ${check.message ?? ''}`);
          }

          // Take screenshot
          await new Promise(r => setTimeout(r, 300));
          const screenshotResp = await kernel.browsers.computer.captureScreenshot(sessionId);
          const buf = Buffer.from(await screenshotResp.arrayBuffer());
          const screenshot = buf.toString('base64');

          items.push({
            type: 'computer_call_output',
            call_id: cc.call_id,
            acknowledged_safety_checks: pending,
            output: {
              type: 'computer_screenshot',
              image_url: `data:image/png;base64,${screenshot}`,
            },
          } as unknown as ResponseItem);
        }
      }

      // Check if the model produced a final assistant message
      const lastItem = response.output[response.output.length - 1] as ResponseItem & { role?: string };
      if (lastItem?.role === 'assistant') {
        const msg = lastItem as ResponseOutputMessage;
        const text = msg.content
          ?.filter(c => c && 'text' in c)
          .map(c => (c as { text: string }).text)
          .join('') ?? '';
        return { result: text || '(no response)', provider: this.name };
      }
    }

    return { result: '(max turns reached)', provider: this.name };
  }
}

File: pkg/templates/python/cua/providers/openai.py
"""OpenAI CUA provider adapter using the Responses API."""

from __future__ import annotations

import asyncio
import base64
import json
import os
from datetime import datetime

import httpx

from . import CuaProvider, TaskOptions, TaskResult

KEYSYM_MAP = {
    "ENTER": "Return", "Enter": "Return", "RETURN": "Return",
    "BACKSPACE": "BackSpace", "Backspace": "BackSpace",
    "DELETE": "Delete", "TAB": "Tab", "ESCAPE": "Escape", "Escape": "Escape",
    "SPACE": "space", "Space": "space",
    "UP": "Up", "DOWN": "Down", "LEFT": "Left", "RIGHT": "Right",
    "HOME": "Home", "END": "End",
    "PAGEUP": "Prior", "PAGE_UP": "Prior", "PageUp": "Prior",
    "PAGEDOWN": "Next", "PAGE_DOWN": "Next", "PageDown": "Next",
    "CTRL": "Control_L", "Ctrl": "Control_L", "CONTROL": "Control_L", "Control": "Control_L",
    "ALT": "Alt_L", "Alt": "Alt_L",
    "SHIFT": "Shift_L", "Shift": "Shift_L",
    "META": "Super_L", "Meta": "Super_L", "CMD": "Super_L", "COMMAND": "Super_L",
}

MODIFIER_KEYSYMS = {
    "Control_L", "Control_R", "Alt_L", "Alt_R",
    "Shift_L", "Shift_R", "Super_L", "Super_R",
}


def _translate_keys(keys: list[str]) -> list[str]:
    return [KEYSYM_MAP.get(k, k) for k in keys]


def _expand_and_translate(
    keys: list[str], hold_keys: list[str],
) -> tuple[list[str], list[str]]:
    expanded = []
    for raw in keys:
        for part in (raw.split("+") if "+" in raw else [raw]):
            trimmed = part.strip()
            if trimmed:
                expanded.append(trimmed)

    translated = _translate_keys(expanded)
    translated_hold = _translate_keys(hold_keys)

    hold_from_keys = [k for k in translated if k in MODIFIER_KEYSYMS]
    primary = [k for k in translated if k not in MODIFIER_KEYSYMS]

    if not primary:
        return translated, translated_hold

    merged = list(dict.fromkeys(translated_hold + hold_from_keys))
    return primary, merged


def _translate_action(action: dict) -> list[dict]:
    action_type = action.get("type", "")

    if action_type == "click":
        button = action.get("button", "left")
        if button == "back":
            return [{"type": "press_key", "press_key": {"keys": ["Left"], "hold_keys": ["Alt_L"]}}]
        if button == "forward":
            return [{"type": "press_key", "press_key": {"keys": ["Right"], "hold_keys": ["Alt_L"]}}]
        if button == "wheel":
            return [{"type": "scroll", "scroll": {
                "x": action.get("x", 0), "y": action.get("y", 0),
                "delta_x": action.get("scroll_x", 0), "delta_y": action.get("scroll_y", 0),
            }}]
        btn = "left"
        if isinstance(button, int):
            btn = {2: "middle", 3: "right"}.get(button, "left")
        elif isinstance(button, str):
            btn = button
        return [{"type": "click_mouse", "click_mouse": {"x": action.get("x", 0), "y": action.get("y", 0), "button": btn}}]

    if action_type == "double_click":
        return [{"type": "click_mouse", "click_mouse": {"x": action.get("x", 0), "y": action.get("y", 0), "num_clicks": 2}}]

    if action_type == "type":
        return [{"type": "type_text", "type_text": {"text": action.get("text", "")}}]

    if action_type == "keypress":
        primary, hold = _expand_and_translate(action.get("keys", []), action.get("hold_keys", []))
        result: dict = {"type": "press_key", "press_key": {"keys": primary}}
        if hold:
            result["press_key"]["hold_keys"] = hold
        return [result]

    if action_type == "scroll":
        return [{"type": "scroll", "scroll": {
            "x": action.get("x", 0), "y": action.get("y", 0),
            "delta_x": action.get("scroll_x", 0), "delta_y": action.get("scroll_y", 0),
        }}]

    if action_type == "move":
        return [{"type": "move_mouse", "move_mouse": {"x": action.get("x", 0), "y": action.get("y", 0)}}]

    if action_type == "drag":
        path = action.get("path", [])
        points = []
        for p in path:
            if isinstance(p, dict):
                points.append([p["x"], p["y"]])
            elif isinstance(p, (list, tuple)) and len(p) >= 2:
                points.append([p[0], p[1]])
        if len(points) < 2:
            raise ValueError("drag requires at least 2 path points")
        return [{"type": "drag_mouse", "drag_mouse": {"path": points}}]

    if action_type == "wait":
        return [{"type": "sleep", "sleep": {"duration_ms": action.get("ms", 1000)}}]

    if action_type == "goto":
        url = action.get("url", "")
        return [
            {"type": "press_key", "press_key": {"keys": ["l"], "hold_keys": ["Control_L"]}},
            {"type": "sleep", "sleep": {"duration_ms": 200}},
            {"type": "press_key", "press_key": {"keys": ["a"], "hold_keys": ["Control_L"]}},
            {"type": "type_text", "type_text": {"text": url}},
            {"type": "press_key", "press_key": {"keys": ["Return"]}},
        ]

    if action_type == "back":
        return [{"type": "press_key", "press_key": {"keys": ["Left"], "hold_keys": ["Alt_L"]}}]

    if action_type == "screenshot":
        return []

    raise ValueError(f"Unknown CUA action: {action_type}")


async def _create_response(api_key: str, **kwargs) -> dict:
    """Call the OpenAI Responses API with retry."""
    async with httpx.AsyncClient(timeout=120) as client:
        for attempt in range(4):
            try:
                resp = await client.post(
                    "https://api.openai.com/v1/responses",
                    headers={
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json",
                    },
                    json=kwargs,
                )
                resp.raise_for_status()
                return resp.json()
            except httpx.HTTPStatusError as exc:
                if exc.response.status_code >= 500 and attempt < 3:
                    await asyncio.sleep(2 ** attempt)
                    continue
                raise
    raise RuntimeError("Max retries exceeded")


class OpenAIProvider:
    name = "openai"

    def __init__(self) -> None:
        self._api_key = os.environ.get("OPENAI_API_KEY", "")

    def is_configured(self) -> bool:
        return len(self._api_key) > 0

    async def run_task(self, options: TaskOptions) -> TaskResult:
        computer = options.kernel.browsers.computer

        # Navigate to starting page
        goto_actions = _translate_action({"type": "goto", "url": "https://duckduckgo.com"})
        await asyncio.to_thread(
            computer.batch, options.session_id, actions=goto_actions,
        )

        input_items = [
            {
                "role": "system",
                "content": (
                    f"Current date: {datetime.now().isoformat()}. "
                    "For long pages, prefer PageUp/PageDown style scrolling over repeated "
                    "mouse-wheel scrolling. Use wheel scrolling mainly for small adjustments."
                ),
            },
            {"type": "message", "role": "user", "content": [{"type": "input_text", "text": options.query}]},
        ]
        items: list[dict] = []

        for _turn in range(50):
            response = await _create_response(
                self._api_key,
                model=options.model or "gpt-5.4",
                input=input_items + items,
                tools=[{"type": "computer"}],
                truncation="auto",
                reasoning={"effort": "low", "summary": "concise"},
            )

            output = response.get("output", [])
            if not output:
                raise RuntimeError("No output from model")

            for item in output:
                items.append(item)

                if item.get("type") == "computer_call":
                    action_list = item.get("actions") or ([item["action"]] if "action" in item else [])

                    batch: list[dict] = []
                    for a in action_list:
                        batch.extend(_translate_action(a))
                    if batch:
                        await asyncio.to_thread(
                            computer.batch, options.session_id, actions=batch,
                        )

                    # Safety checks
                    for check in item.get("pending_safety_checks", []):
                        print(f"Safety check: {check.get('message', '')}")

                    await asyncio.sleep(0.3)
                    resp = await asyncio.to_thread(
                        computer.capture_screenshot, options.session_id,
                    )
                    screenshot = base64.b64encode(resp.read()).decode()

                    items.append({
                        "type": "computer_call_output",
                        "call_id": item["call_id"],
                        "acknowledged_safety_checks": item.get("pending_safety_checks", []),
                        "output": {
                            "type": "computer_screenshot",
                            "image_url": f"data:image/png;base64,{screenshot}",
                        },
                    })

            # Check for final assistant message
            last = output[-1] if output else {}
            if last.get("role") == "assistant":
                content = last.get("content", [])
                texts = [c.get("text", "") for c in content if isinstance(c, dict) and "text" in c]
                return TaskResult(result=" ".join(texts) or "(no response)", provider=self.name)

        return TaskResult(result="(max turns reached)", provider=self.name)
